### PR TITLE
Optimize GLM-5.3 B12X and DFlash2 MXFP8 serving

### PR DESCRIPTION
**Pull request by voipmonitor**
## Resulting behavior

Status: **qualified** for GLM-5.3-Flash NVFP4 serving at tensor parallel size 4 with a ModelOpt MXFP8 DFlash2 drafter.

- GLM C4 decode uses the B12X paged selector. GLM C4 prefill gathers the visible packed cache into caller-owned workspace and uses DeepGEMM for ranking.
- Multipath hyperconnection uses B12X for decode-sized batches and TileLang for prefill-sized batches.
- GLM KDA backend selection is fixed for the server lifetime. Automatic selection uses Triton without speculation and B12X with MTP or DFlash2. An operator can explicitly select either backend.
- Full CUDA graph capture retains Python resource owners whose device pointers are embedded in captured custom operations.
- DFlash2 applies serialized ModelOpt MXFP8 quantization to convolution, candidate-selector, and fused context K/V projections. The context path executes one fused K/V GEMM and does not compute discarded Q rows.
- Target and draft attention groups retain independent KV-cache dtypes while sharing the engine-selected physical cache layout.
- ModelConfig.revision controls GLM processor, tokenizer, image-processor, and video-processor artifacts, preventing pinned weights from mixing with artifacts from a moving repository branch.
- GLM sparse-cache tests assert the QSA page-tail ABI consumed by B12X.

## Source contract

- vLLM base: local-inference-lab/vllm dev/jovian-judgement at 015dcd423d6aabf843c8ad69074ff67d35c2a395.
- Pull-request head: b77333ca8824897ff6ddf96a62208ea406c555a9.
- Result tree: 82bbab85cebf48b735b5898062ceca89826b3913.
- Qualified B12X runtime base: 2fcf23a0ce269be27b2e03fece73d46e90e6aeea.
- B12X graph-replay qualification: [B12X PR #250](https://github.com/local-inference-lab/b12x/pull/250) at dd8cf60505e0363ab9d6ef6b2116c3a37216a2f1, result tree fdbb504ccf5842ae7bb7089caa01fc076a7043c0. That pull request changes tests only.

The source-locked image recipe is [blackwell-llm-docker PR #28](https://github.com/local-inference-lab/blackwell-llm-docker/pull/28). It reconstructs both result trees from immutable patches and validates their SHA-256 hashes without applying any additional vLLM or B12X source patch.

## Compatibility and limitations

- Target checkpoint: local-inference-lab/GLM-5.3-Flash-NVFP4 at 520de24eabf507659eaef7c70f14fd584527facc.
- DFlash2 checkpoint: local-inference-lab/GLM-5.3-Flash-DFlash2-MXFP8 at b6d33aa93fc1ac5b23a88251a1c0ce0bfe2ad17c, with seven proposed tokens.
- Unquantized DFlash checkpoints retain the fused BF16 context projection. The quantized projection path is selected only for serialized ModelOpt MXFP8.
- Mixed MXFP8 and non-MXFP8 context projection layers are rejected at model construction in either layer order.
- Target-only attention groups and deployments using one global cache configuration retain their established behavior.
- The target GDN backend captures decode. Unsupported target prefill segments execute eagerly. DFlash2 graph families capture in full mode.
- DFlash2 causal attention uses FlashAttention 2 because B12X does not implement the draft causal-attention contract.

## Validation

- Focused attention-group, worker, and DFlash2 MXFP8 auxiliary-linear tests: 24 passed.
- GLM model, pooled-indexer, sparse-attention, and workspace suites: 80 passed; 14 accelerator-dependent tests skipped.
- Review regression tests: mixed-quantization rejection 2 passed; breakable graph resource retention 2 passed; CPU breakable graph file 4 passed and 11 CUDA tests skipped; workspace 8 passed; attention-group head counts 2 passed and 4 accelerator tests skipped.
- A physical SM120 GPU oracle compared DeepGEMM and B12X ranking over 576 valid candidates at top-k 512: passed.
- Ruff check and format, mypy, SPDX, forbidden-import, configuration, typo, and all other applicable commit hooks: passed.
- Source-locked Docker composition and audit tests: 17 passed.
- Published image: voipmonitor/vllm:glm53-flash-dflash2-mxfp8-vllm82bbab8-b12xfdbb504-fi1ac6942-cu133-torch213-20260828-r1-gitd3920aaa7c05.
- Published manifest digest: sha256:1a210a6dcd4eeef4b9515aa03b8117dbd1bad70ed101cdf72ec4692015e2ac4b.
- TP4 serving on physical GPUs 4, 5, 6, and 7 loaded both pinned checkpoint revisions, initialized B12X PCIe all-reduce, selected B12X NVFP4 target MoE and B12X MXFP8 draft linear kernels, selected FlashAttention 2 for draft causal attention, captured target decode and all 16 DFlash2 full-graph shapes, passed health and model-list requests, and returned HTTP 200 for an OpenAI-compatible completion.
- CC16 decode on the published image for 30 seconds: 987.69 output tokens/s, 363.76 verifier steps/s, effective accepted length 2.7153, and zero request errors.
- Standalone 32,768-token cold-prefill profile on the published image for 30 seconds: 12,692 input tokens/s over 10 samples.

## Duplicate-work check

[vLLM PR #490](https://github.com/local-inference-lab/vllm/pull/490) integrates a proposed B12X direct single-token KDA API for non-speculative decode. It does not cover speculative KDA transactions, C4 prefill, MXFP8 DFlash2 projections, independent target and draft cache types, or graph resource retention.

[vLLM PR #488](https://github.com/local-inference-lab/vllm/pull/488) targets decode-context-parallel CKV gathering and DFlash cache retention. This pull request qualifies tensor parallel size 4 with decode context parallel size 1 and does not import the DCP-specific cache coordinator.

Upstream draft [vllm-project/vllm#51620](https://github.com/vllm-project/vllm/pull/51620) addresses the same raw-weight failure with a generic per-layer quantized fallback. The implementation here combines ModelOpt MXFP8 values and E8M0 scales into one fused projection and dispatches through the selected quantized linear kernel, avoiding per-layer QKV execution and discarded Q output.

## Review disclosure

OpenAI Codex assisted with implementation, debugging, tests, profiling, benchmarking, commit preparation, and pull-request text. Human review of every changed line and the serving contract is required before merge.

### CODE REVIEW

File: vllm/model_executor/models/qwen3_dflash.py [hunk 1]
```diff
@@ -491,6 +499,25 @@ def _build_context_kv_buffers(
...
+                for a in layers_attn
+            ):
+                raise ValueError(
+                    "Every DFlash attention layer must use the same MXFP8 "
+                    "linear format for the fused context K/V projection."
+                )
+            kv_scales = [a.qkv_proj.weight_scale[a.q_size :] for a in layers_attn]
+            self._fused_kv_weight_scale = torch.cat(kv_scales, dim=0)
+        else:
+            self._fused_kv_weight_scale = None
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply vllm/model_executor/models/qwen3_dflash.py]: Fixed in b77333ca882. The DFlash context projection classifies every layer quantization method before reading any serialized weight, rejects mixed MXFP8 and non-MXFP8 configurations in either order, and initializes block scales only for a uniformly MXFP8 layer set. The two-order regression test passed.

File: vllm/v1/worker/gpu/cudagraph_utils.py [hunk 2]
```diff
@@ -377,18 +379,20 @@ def capture(
...
-                            # the next forward pass.
+                            # Join the offloader copy stream because the last layer
+                            # can leave a prefetch pending at capture end.
                             get_offloader().join_after_forward()
                         if self._capture_mem_samples is not None:
                             torch.accelerator.synchronize()
                             free_after = torch.accelerator.get_memory_info()[0]
                             self._capture_mem_samples.append(free_before - free_after)
                         self.graphs[desc] = graph
+                        self.graph_capture_resources[desc] = resources
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply vllm/v1/worker/gpu/cudagraph_utils.py:395]: Fixed in b77333ca882. BreakableCUDAGraphWrapper now opens a capture resource collector around the complete breakable capture and stores the retained owners on the corresponding graph entry. B12X bindings therefore remain alive for the same lifetime as their graph entry. The resource-lifetime regression test passed, and the breakable test file reported 4 passed with 11 CUDA-only tests skipped in the CPU run.

File: vllm/v1/worker/workspace.py [hunk 3]
```diff
@@ -44,6 +48,33 @@ def use_workspace_lane(lane: int) -> Iterator[None]:
...
+        _cuda_graph_capture_resources.reset(token)
+
+
+def retain_cuda_graph_capture_resource(resource: Any) -> bool:
+    """Retain ``resource`` when called inside full CUDA graph capture."""
+    resources = _cuda_graph_capture_resources.get()
+    if resources is None:
+        return False
+    resources.append(resource)
+    return True
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply vllm/v1/worker/workspace.py:82]: Fixed in b77333ca882. Both retain_cuda_graph_capture_resource and get_kv_cache_dtype_from_layers now document their arguments and return contracts with Google-style sections. Ruff and repository commit hooks passed.

**Comment by voipmonitor** [comment]
Review follow-up in b77333ca882 also removes the positional common_args[10] splice from the GLM C4 prefill kernel launch. The DeepGEMM/B12X prefill oracle now uses 576 valid pooled-cache candidates with the production-supported top-k of 512, so the comparison exercises score ranking instead of only valid-range and padding behavior. That oracle passed on physical NVIDIA RTX PRO 6000 Blackwell GPU 7 while the TP4 serving process remained active.

**Comment by voipmonitor** [comment]
Review status: the source changes are partitioned into independently reviewable pull requests:

- #493 retains custom-operation resources for CUDA graph lifetime.
- #494 preserves per-attention-group KV cache formats.
- #495 defines caller-scratch KDA binding and server-lifetime backend selection.
- #496 implements GLM C4 packed-cache prefill and parallel pool updates.
- #497 pins GLM processor artifacts to ModelConfig.revision.
- #498 dispatches GLM multipath hyperconnection by rank-local token count.
- #499 executes serialized ModelOpt MXFP8 DFlash2 projections.

All seven branches use dev/jovian-judgement commit c79f35ca00e8e93e0943a0d79b85b22b18aac939 as their base. Merging their heads produces tree 64c2eb4eb03a64660dc2965a3164004a5e9cec0f, which exactly matches the Git merge tree of that base and this pull request at b77333ca8824897ff6ddf96a62208ea406c555a9.

The branch for #491 remains available as the immutable source snapshot used by blackwell-llm-docker #28 and the published image with manifest digest sha256:1a210a6dcd4eeef4b9515aa03b8117dbd1bad70ed101cdf72ec4692015e2ac4b. No source or qualification evidence is deleted. B12X #250 remains a separate test-only qualification pull request.
